### PR TITLE
Fix API raw requests for commits and tags

### LINKED ISSUE #2824: Integration between drone and Gitea (bleeding edge) erroring out on webhook
- Gitea version (or commit ref): e6a97b6
- Git version: 2.13.5
- Operating system: centos (Docker)
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
## Description
When I push a commit to a repo that has a webhook that is connected to drone the webhook fails. Checking the logs from gitea I see that drone fetches a URL from the API, but is returned a 404 (not found) error.
I've tested calling that route directly via browser and I also get a 404. However when I call /api/v1/repos/sdstack/atomic/ directly via my browser (when I am logged in) I get information about the repo, and when I am logged out (testing via incognito mode) I call the /api/v1/repos/sdstack/atomic/ URL and it gives me an expected 403 (as it is a private repo so I wouldn't want someone not logged in to be able to get that information). 

Drone version: 0.8.1+build.1318

172.18.0.100 - - [01/Nov/2017:20:39:28 +0300] "GET /api/v1/repos/sdstack/atomic/raw/6d31739044602484ef3b658270a973c5ebedb461/.drone.yml HTTP/1.1" 404 11349 "-" "Go-http-client/1.1" "-"

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #2824

Replaces #2836

### CODE REVIEW

File: modules/context/repo.go [hunk 1]
```diff
@@ -497,6 +500,8 @@ func getRefNameFromPath(ctx *Context, path string, isExist func(string) bool) st
 func getRefName(ctx *Context, pathType RepoRefType) string {
 	path := ctx.Params("*")
 	switch pathType {
+	case RepoRefAPI:
+		fallthrough
```
Comment (Morlinest) [inline modules/context/repo.go]: Is not this same as `case RepoRefLegacy, RepoRefAPI:`? That means you don't need `RepoRefAPI`.

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by appleboy** [comment]
Still failed in push tag event.

```
ERRO[0095] error: fullinn/back: cannot find .drone.yml in refs/tags/0.0.10: 404 Not Found
ERRO[0095] Error #01: 404 Not Found
                     ip=139.162.127.189 latency=28.606842ms method=POST path=/hook status=404 time=2017-11-03T14:36:50Z user-agent=GogsServer
```

File: modules/context/repo.go (hunk 1 above)
Comment (ethantkoenig) [reply modules/context/repo.go]: Yes, you do; see [line 623](https://github.com/go-gitea/gitea/pull/2841/files#diff-63a3dd20bb1fd6e9042a6f3353965a58R623)

Comment (Morlinest) [reply modules/context/repo.go]: @ethantkoenig Thx, got it. `case RepoRefLegacy, RepoRefAPI:` can be used anyway.

File: modules/context/repo.go [hunk 2]
```diff
@@ -466,6 +466,9 @@ const (
 	// RepoRefLegacy unknown type, make educated guess and redirect.
 	// for backward compatibility with previous URL scheme
 	RepoRefLegacy RepoRefType = iota
+	// RepoRefAPI is for usage in API where educated guess is needed
+	// but redirect can not be made
+	RepoRefAPI
```
Comment (ethantkoenig) [inline modules/context/repo.go]: nit: rename to make behavior more obvious, perhaps `RepoRefAny`?

**Comment by lafriks** [comment]
@Morlinest @ethantkoenig fixed your recommendations

**Comment by Morlinest** [comment]
LGTM

**Review by Morlinest — APPROVED** [review]

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Comment by lafriks** [comment]
@appleboy it does work for me (at least when accessing API directly). Added also tests that show that

**Comment by lafriks** [comment]
@appleboy can you please verify from logs that api/raw url is what is failing?

**Comment by appleboy** [comment]
@lafriks Still not working in Tag event.

**Pull request merged**